flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -525,63 +525,88 @@ class ExprLowering {
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        } else if constexpr (KIND == 16) {
-          llvm::APFloat floatVal{llvm::APFloatBase::IEEEquad(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        } else {
-          // convert everything else to double
-          llvm::APFloat floatVal{llvm::APFloatBase::IEEEdouble(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        }
+      std::string str = value.DumpHexadecimal();
+      if constexpr (KIND == 2) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEhalf(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 4) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEsingle(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 10) {
+        llvm::APFloat floatVal{llvm::APFloatBase::x87DoubleExtended(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 16) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEquad(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else {
+        // convert everything else to double
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEdouble(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
       }
-      llvm_unreachable("real constant has no value");
     } else if constexpr (TC == Fortran::lower::ComplexCat) {
-      auto opt = con.GetScalarValue();
-      if (opt.has_value()) {
-        using TR = Fortran::evaluate::Type<Fortran::lower::RealCat, KIND>;
-        return genval(Fortran::evaluate::ComplexConstructor<KIND>{
-            Fortran::evaluate::Expr<TR>{
-                Fortran::evaluate::Constant<TR>{opt->REAL()}},
-            Fortran::evaluate::Expr<TR>{
-                Fortran::evaluate::Constant<TR>{opt->AIMAG()}}});
-      }
-      llvm_unreachable("array of complex unhandled");
-    } else if constexpr (TC == Fortran::lower::CharacterCat) {
-      return genCharLit<KIND>(con.GetScalarValue().value(), con.LEN());
+      using TR = Fortran::evaluate::Type<Fortran::lower::RealCat, KIND>;
+      return genval(Fortran::evaluate::ComplexConstructor<KIND>{
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.REAL()}},
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.AIMAG()}}});
     } else {
       llvm_unreachable("unhandled constant");
     }
   }
 
+  template <Fortran::common::TypeCategory TC, int KIND>
+  mlir::Value genArrayLit(
+      const Fortran::evaluate::Constant<Fortran::evaluate::Type<TC, KIND>>
+          &con) {
+    // TODO:
+    // - Multi-dimensional array
+    if (con.Rank() > 1) {
+      TODO(); // "Multi dimensional array not yet supported"
```

Comment:
Does TODO take a message so we can give the user a hint about what is NYI?

Reply:
Yes, also guides us developers on what needs to be done.